File: packages/testing/glm/weighted-glm-vcov.test.R
#!/usr/bin/env Rscript

# Test weighted GLM with vcov() and confint()
# This checks if edge cases affect variance-covariance and confidence intervals

cat("=== Weighted GLM vcov/confint Edge Case Tests ===\n\n")

# Test 1: Normal weighted GLM - baseline
cat("Test 1: Normal weighted GLM - baseline\n")
x <- c(1, 2, 3, 4, 5)
y <- c(2.1, 4.2, 5.8, 8.1, 10.3)
weights <- c(1, 1, 2, 2, 1)
model1 <- glm(y ~ x, family = gaussian(link = "identity"), weights = weights)
cat("Coefficients:\n")
print(coef(model1))
cat("\nVcov matrix:\n")
print(vcov(model1))
cat("\nConfidence intervals (95%):\n")
print(confint.default(model1, level = 0.95))
cat("\n")

# Test 2: Single non-zero weight - edge case
cat("Test 2: Single non-zero weight - edge case\n")
weights2 <- c(0, 0, 0, 0, 1)
model2 <- glm(y ~ x, family = gaussian(link = "identity"), weights = weights2)
cat("Coefficients:\n")
print(coef(model2))
cat("\nVcov matrix:\n")
print(vcov(model2))
cat("\nNote: Slope is NA because only one observation\n")
cat("\n")

# Test 3: Identical x values - edge case
cat("Test 3: Identical x values with different weights - edge case\n")
x3 <- c(1, 1, 1, 1, 1)
y3 <- c(2.1, 4.2, 5.8, 8.1, 10.3)
weights3 <- c(1, 2, 3, 4, 5)
model3 <- glm(y3 ~ x3, family = gaussian(link = "identity"), weights = weights3)
cat("Coefficients:\n")
print(coef(model3))
cat("\nVcov matrix:\n")
print(vcov(model3))
cat("\nNote: Slope is NA because all x values are identical\n")
cat("\n")

# Test 4: Single observation
cat("Test 4: Single observation\n")
x4 <- c(1)
y4 <- c(2.1)
weights4 <- c(1)
model4 <- glm(y4 ~ x4, family = gaussian(link = "identity"), weights = weights4)
cat("Coefficients:\n")
print(coef(model4))
cat("\nVcov matrix:\n")
print(vcov(model4))
cat("\nNote: Slope is NA because only one observation\n")
cat("\n")

# Test 5: Very small weights
cat("Test 5: Very small weights\n")
weights5 <- c(1e-10, 1e-8, 1e-6, 1e-4, 1e-2)
model5 <- glm(y ~ x, family = gaussian(link = "identity"), weights = weights5)
cat("Coefficients:\n")
print(coef(model5))
cat("\nVcov matrix:\n")
print(vcov(model5))
cat("\nConfidence intervals (95%):\n")
print(confint.default(model5, level = 0.95))
cat("\n")

# Test 6: Binomial with weights and vcov
cat("Test 6: Binomial with weights and vcov\n")
successes <- c(8, 12, 15, 18, 20)
trials <- c(10, 15, 20, 25, 30)
x6 <- c(1, 2, 3, 4, 5)
y6 <- successes / trials
model6 <- glm(y6 ~ x6, family = binomial(link = "logit"), weights = trials)
cat("Coefficients:\n")
print(coef(model6))
cat("\nVcov matrix:\n")
print(vcov(model6))
cat("\nConfidence intervals (95%):\n")
print(confint.default(model6, level = 0.95))
cat("\n")

cat("=== All Tests Complete ===\n")

File: packages/testing/glm/weighted-glm-vcov.test.ts
import { glm } from "../../dataframe/ts/wasm/glm-functions.ts";
import { createDataFrame } from "@tidy-ts/dataframe";
import { assertClose, TOL } from "./glm-test-helpers.ts";

// Test weighted GLM with vcov() and confint()
// This checks if edge cases affect variance-covariance and confidence intervals

Deno.test("Weighted GLM vcov/confint - Test 1: Normal weighted GLM - baseline", () => {
  const x = [1, 2, 3, 4, 5];
  const y = [2.1, 4.2, 5.8, 8.1, 10.3];
  const weights = [1, 1, 2, 2, 1];

  const df = createDataFrame({ columns: { x, y } });

  const result = glm({
    formula: "y ~ x",
    family: "gaussian",
    link: "identity",
    data: df,
    options: { weights },
  });

  // Check coefficients
  assertClose(result.coefficients[0], -0.04210526, TOL, "weighted coef[0]");
  assertClose(result.coefficients[1], 2.03157895, TOL, "weighted coef[1]");

  // Check vcov matrix
  const vcov = result.vcov();
  assertClose(vcov[0][0], 0.07239151, TOL, "weighted vcov[0][0]");
  assertClose(vcov[0][1], -0.019907664, TOL, "weighted vcov[0][1]");
  assertClose(vcov[1][0], -0.01990766, TOL, "weighted vcov[1][0]");
  assertClose(vcov[1][1], 0.006334257, TOL, "weighted vcov[1][1]");

  // Check confidence intervals
  const ci = result.confint({ level: 0.95 });
  assertClose(ci.lower[0], -0.5694467, TOL, "weighted CI lower[0]");
  assertClose(ci.upper[0], 0.4852362, TOL, "weighted CI upper[0]");
  assertClose(ci.lower[1], 1.8755892, TOL, "weighted CI lower[1]");
  assertClose(ci.upper[1], 2.1875686, TOL, "weighted CI upper[1]");
});

Deno.test("Weighted GLM vcov/confint - Test 2: Single non-zero weight - edge case", () => {
  const x = [1, 2, 3, 4, 5];
  const y = [2.1, 4.2, 5.8, 8.1, 10.3];
  const weights2 = [0, 0, 0, 0, 1];

  const df = createDataFrame({ columns: { x, y } });

  const result = glm({
    formula: "y ~ x",
    family: "gaussian",
    link: "identity",
    data: df,
    options: { weights: weights2 },
  });

  // R gives: Intercept=10.3, Slope=NA
  // TS currently gives: Intercept=-10.3, Slope=0
  console.log("Single non-zero weight:");
  console.log("  TS Intercept:", result.coefficients[0], "R: 10.3");
  console.log("  TS Slope:", result.coefficients[1], "R: NA");

  // Check vcov - R gives NaN for intercept, NA for others
  const vcov = result.vcov();
  console.log("  TS vcov[0][0]:", vcov[0][0], "R: NaN");
  console.log("  TS vcov[1][1]:", vcov[1][1], "R: NA");
});

Deno.test("Weighted GLM vcov/confint - Test 3: Identical x values - edge case", () => {
  const x3 = [1, 1, 1, 1, 1];
  const y3 = [2.1, 4.2, 5.8, 8.1, 10.3];
  const weights3 = [1, 2, 3, 4, 5];

  const df = createDataFrame({ columns: { x: x3, y: y3 } });

  const result = glm({
    formula: "y ~ x",
    family: "gaussian",
    link: "identity",
    data: df,
    options: { weights: weights3 },
  });

  // R gives: Intercept=7.453333, Slope=NA
  // TS currently gives: Intercept=0, Slope=0
  console.log("Identical x values:");
  console.log("  TS Intercept:", result.coefficients[0], "R: 7.453333");
  console.log("  TS Slope:", result.coefficients[1], "R: NA");

  // Check vcov - R gives 1.670289 for intercept variance, NA for slope
  const vcov = result.vcov();
  console.log("  TS vcov[0][0]:", vcov[0][0], "R: 1.670289");
  console.log("  TS vcov[1][1]:", vcov[1][1], "R: NA");
});

Deno.test("Weighted GLM vcov/confint - Test 4: Single observation", () => {
  const x4 = [1];
  const y4 = [2.1];
  const weights4 = [1];

  const df = createDataFrame({ columns: { x: x4, y: y4 } });

  const result = glm({
    formula: "y ~ x",
    family: "gaussian",
    link: "identity",
    data: df,
    options: { weights: weights4 },
  });

  // R gives: Intercept=2.1, Slope=NA
  // TS currently gives: Intercept=-2.1, Slope=0
  console.log("Single observation:");
  console.log("  TS Intercept:", result.coefficients[0], "R: 2.1");
  console.log("  TS Slope:", result.coefficients[1], "R: NA");

  // Check vcov - R gives NaN for intercept, NA for slope
  const vcov = result.vcov();
  console.log("  TS vcov[0][0]:", vcov[0][0], "R: NaN");
  console.log("  TS vcov[1][1]:", vcov[1][1], "R: NA");
});

Deno.test("Weighted GLM vcov/confint - Test 5: Very small weights", () => {
  const x = [1, 2, 3, 4, 5];
  const y = [2.1, 4.2, 5.8, 8.1, 10.3];
  const weights5 = [1e-10, 1e-8, 1e-6, 1e-4, 1e-2];

  const df = createDataFrame({ columns: { x, y } });

  const result = glm({
    formula: "y ~ x",
    family: "gaussian",
    link: "identity",
    data: df,
    options: { weights: weights5 },
  });

  // Check coefficients
  assertClose(result.coefficients[0], -0.7089085, TOL, "small-weights coef[0]");
  assertClose(result.coefficients[1], 2.2017834, TOL, "small-weights coef[1]");

  // Check vcov matrix
  const vcov = result.vcov();
  assertClose(vcov[0][0], 0.0009834810, TOL, "small-weights vcov[0][0]");
  assertClose(vcov[0][1], -1.970136e-04, TOL, "small-weights vcov[0][1]");
  assertClose(vcov[1][0], -0.0001970136, TOL, "small-weights vcov[1][0]");
  assertClose(vcov[1][1], 3.948249e-05, TOL, "small-weights vcov[1][1]");

  // Check confidence intervals
  const ci = result.confint({ level: 0.95 });
  assertClose(ci.lower[0], -0.7703739, TOL, "small-weights CI lower[0]");
  assertClose(ci.upper[0], -0.647443, TOL, "small-weights CI upper[0]");
  assertClose(ci.lower[1], 2.1894680, TOL, "small-weights CI lower[1]");
  assertClose(ci.upper[1], 2.214099, TOL, "small-weights CI upper[1]");
});

Deno.test("Weighted GLM vcov/confint - Test 6: Binomial with weights and vcov", () => {
  const successes = [8, 12, 15, 18, 20];
  const trials = [10, 15, 20, 25, 30];
  const x6 = [1, 2, 3, 4, 5];
  const y6 = successes.map((s, i) => s / trials[i]);

  const df = createDataFrame({ columns: { x: x6, y: y6 } });

  const result = glm({
    formula: "y ~ x",
    family: "binomial",
    link: "logit",
    data: df,
    options: { weights: trials },
  });

  // Check coefficients
  assertClose(result.coefficients[0], 1.6953808, TOL, "binomial-weighted coef[0]");
  assertClose(result.coefficients[1], -0.1957673, TOL, "binomial-weighted coef[1]");

  // Check vcov matrix
  const vcov = result.vcov();
  assertClose(vcov[0][0], 0.4752707, TOL, "binomial-weighted vcov[0][0]");
  assertClose(vcov[0][1], -0.11584094, TOL, "binomial-weighted vcov[0][1]");
  assertClose(vcov[1][0], -0.1158409, TOL, "binomial-weighted vcov[1][0]");
  assertClose(vcov[1][1], 0.03165606, TOL, "binomial-weighted vcov[1][1]");

  // Check confidence intervals (profile CI - R's confint())
  const ci = result.confint({ level: 0.95 });
  assertClose(ci.lower[0], 0.4183132, 1e-4, "binomial-weighted CI lower[0]");
  assertClose(ci.upper[0], 3.150958, 1e-4, "binomial-weighted CI upper[0]");
  assertClose(ci.lower[1], -0.5601602, 1e-4, "binomial-weighted CI lower[1]");
  assertClose(ci.upper[1], 0.1436985, 1e-4, "binomial-weighted CI upper[1]");
});
